File: singleton-pattern/java/src/SingleObject.java
package src;

public class SingleObject {
  // 类加载时就初始化，浪费内存
  private static final SingleObject singleObject = new SingleObject();

  // 构造函数是private，不允许实例化
  private SingleObject() {

  }
  public static SingleObject getInstance() {
    return singleObject;
  }

  public void run() {
    System.out.println("SingleObject::run()");
  }
}

File: singleton-pattern/ts/src/SingleObject.ts
export const SingleObject = (function () {
  let instance: any = undefined

  // 对象创建方式，如果已经实例化则直接返回对象，类似一个普通Object
  function createInstance(alias: string) {
    if (instance) {
      return instance
    }
    instance = Object.create(null)
    instance.alias = alias
    instance.run = function () {
      console.log('SingleObject::()', instance.alias)
    }
    return instance
  }

  return {
    getInstance: function (alias: string) {
      if (!instance) {
        instance = createInstance(alias)
      }
      return instance
    }
  }
})()
